### PR TITLE
[main] Add new method to track originating cluster of Backup

### PR DESCRIPTION
**Pull request by mallardduck**
This partial solves for https://github.com/rancher/backup-restore-operator/issues/612

However it's important to note that while I wrote an RFE for this too - the root here is seeking to fix a bug. The bug is that "BRO itself doesn't do anything to give indications when In-Place Restore is acceptable to use".

The short version of the fix, is we add a few new status details and expose them via `kubectl`. And later make a change to `rancher/dashboard` to ensure UI can also benefit from this new context by adding a column.

---

To be discussed:

- This change is very Backwards Compatible, can we ship a CRD change for the new status field in a `2.10.x` release?
  - Feels like it's a grey area - but, probably OK.
  - `k8s` seems to be OK with full on schema additions (of `spec` too) as long as they're optional/nullalble. This is hopefully less impactful since users cannot control `status`.
  - The current result on a cluster w/o CRD change is:
    -  is a new warning about `originCluster` not existing on status,
    - Status conditions are correctly set to represent unknown in-place viability.
- Do we want to explore supporting an even more Backwards friendly method of detecting origin ID's with Filename?
  - For the best UX, we will account for existing backups with UID in filename. In other words, once the CRD is updated with new status field we'll add the UID we find from the filename (or do nothing if not matching expected filename format).

### CODE REVIEW

**Comment by mallardduck** [comment]
@jbiers - So i'm not moving it to in review yet, since this was a very rough first pass try. However if you can take some time to get familiar with the changes and review the list of concerns I thought of (as well as consider/add your own) - that'd be great help to keep the ball rolling on this. In other words, while I could work on this more I'd love your feedback first before I spend more time on it.

File: pkg/util/util.go [hunk 1]
```diff
@@ -74,3 +78,47 @@ func ErrList(e []error) error {
...
+		logrus.Infof("Error fetching CRD: %v", err)
+		return false
+	}
+
+	// Inspect the status schema, for example
+	_, found := crd.Spec.Versions[0].Schema.OpenAPIV3Schema.Properties["status"].Properties["originCluster"]
+	if found {
+		logrus.Debugf("Status schema contains `originCluster` on CRD `%s`.\n", crdName)
+		return true
+	}
```
Comment (alexandreLamarre) [inline pkg/util/util.go:120]: Going back to the above discussion, storing the cluster metadata in the backup CRD spec as a nullable obejct shouldn't be a breaking change, and we can instead check if it is nil or not, rather than doing dynamic CRD reflection

IMO it is easier to make it testable if we avoid checking crd definitions and focus on the presence/absence of a new field to determine that.

File: pkg/controllers/backup/cluster_origin.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	conditionMeta.hasInPlaceRestoreCondition = currentInPlaceRestoreString != ""
+	if !conditionMeta.hasInPlaceRestoreCondition {
+		conditionMeta.currentInPlaceRestoreCondition = currentInPlaceRestoreString == "True"
+	}
+
+	return conditionMeta
+}
+
+// prepareClusterOriginConditions helps set the cluster origin conditions and reports if anything changed in this part of status.
+func (h *handler) prepareClusterOriginConditions(backup *v1.Backup) bool {
```
Comment (alexandreLamarre) [inline pkg/controllers/backup/cluster_origin.go:56]: I think a lot of this logic in this function can also be simplified/more readable if we encapsulate most of it in struct methods

File: pkg/controllers/backup/controller.go [hunk 3]
```diff
@@ -39,7 +39,9 @@ type handler struct {
 	dynamicClient           dynamic.Interface
 	defaultBackupMountPath  string
 	defaultS3BackupLocation *v1.S3ObjectStore
-	kubeSystemNS            string
+	// TODO: rename to kubeSystemNamespaceUID; nit to improve clarity, it's not the string representation nor the NS resource
+	kubeSystemNS              string
+	canUseClusterOriginStatus bool
```
Comment (alexandreLamarre) [inline pkg/controllers/backup/controller.go:44]: stylistically, should avoid storing additional boolean fields, in favor of a method. 

I think this field might be useless?

If we opt into using the backup CRD spec for storing cluster metadata, and older backup-restore versions don't set this new null-able field, then we cannot determine where it is from.

File: pkg/controllers/backup/cluster_origin.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	conditionChanged := false
+	if !h.canUseClusterOriginStatus {
+		currentOriginConditionString := condition.Cond(v1.BackupConditionClusterOrigin).GetStatus(backup)
+		if currentOriginConditionString != "False" {
+			condition.Cond(v1.BackupConditionClusterOrigin).SetStatusBool(backup, false)
+			condition.Cond(v1.BackupConditionClusterOrigin).Message(backup, "CRD not updated to include cluster UID yet.")
```
Comment (alexandreLamarre) [inline pkg/controllers/backup/cluster_origin.go:62]: When would this realistically happen?

**Review by alexandreLamarre — COMMENTED** [review]
Mostly some small comments about making the code more simple,readable & testable.

Since this PR could be a stepping stone to fully allowing backups of backup CRDs themselves, storing the cluster origin metadata in the Status feels wrong.

But if we do intend to fully support taking backups of backup CRDs, then even storing the cluster origin metadata in the Spec section of the CRD feels wrong.

What happens when we have a list of remote backups say `A`, `B`, `C`, all taken from different origins?


If the current cluster matches origin `C` then our status will indicate the restore is viable even if they specify an older backup `A` or `B`. I know this hypothetical would be rare but I think it points to the fact that maybe we should think about this differently.

The current design labels the recurring backups as coming from this cluster only, whereas specific backups are specific to specific clusters

What if we embedded cluster origin metadata not in the CRDs, but the backup data itself? 

This way the flow looks like 
- Each remote backup contains a cluster origin
- Each restore pointing to a specific backup can determine before actually restoring objects if it is viable
-  We can add a flag to the restore spec to allow `unsafe` backups, ones where the cluster origin doesn't match
- By default, we disallow cases where we could have unsafe restores
- If the restore is unsafe, the status is surfaced through the restore CR
- Users can then modify the restore CR to restore the backup even if it is unsafe after the fact, or opt in before creating the restore CR


From the end user's perspective it is somewhat clunkier to take a restore, then be informed that it is potentially unsafe, but as an end user I value being shown correct information more


Additionally storing, the cluster origin metadata in the backup data, allows us to add tests for each corner case in the e2e tests when taking a restore, producing a potentially less volatile new feature

File: pkg/controllers/backup/cluster_origin.go (hunk 4 above)
Comment (mallardduck) [reply pkg/controllers/backup/cluster_origin.go:62]: Yeah, at least I think so. Granted it was during devworkflows, but I did run into it myself.

Maybe not as big a concern in prod, but I think could still happen - in a chart like BRO if they don't also update the CRDs. I.e. Installed via helm CLI directly.
Or even if (an edge case) where a user manually targets a newer image than the chart. However if we opt to not change the CRD to add it to status this wouldn't be needed anyway.

**Comment by jbiers** [comment]
While testing some stuff with our backups I found something we may not have taken into account: if you apply a One-Time backup in a different cluster, the operator will actually pick it up and execute it again. For that reason, it in-place backups will always be possible

**Comment by alexandreLamarre** [comment]
:/

**Comment by jbiers** [comment]
:O
